<a href="https://colab.research.google.com/github/juancpv/predictivemaintenance/blob/master/lstm.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
import keras
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Some tools
from sklearn import preprocessing
from sklearn.metrics import confusion_matrix, recall_score, precision_score
from keras.models import Sequential
from keras.layers import Dense, Dropout, LSTM

Using TensorFlow backend.



  **DATASETS.**

Para este ejercicio nosotros haremos uso de un dataset de Azure. Los datos consisten en tres archivos de texto: training, testing y ground truth en el caso de motores de aeronaves. 


*   El dataset de entrenamiento consiste de un conjunto de series de tiempo multivariable (21 sensores) con la unidad de tiempo medida en ciclos.
*   El dataset de pruebas, tiene el mismo esquema que el de entrenamiento, con la unica diferencia que estos no incluyen la etiqueta de ocurrencia de falla.
*   El datset de ground truth contiene información de ciclos restantes de trabajo de los motores




A continuación se dan los comandos para obtener los datasets (recuerde que esto es una maquina con sistema operativo Linux).





In [1]:
!wget http://azuremlsamples.azureml.net/templatedata/PM_train.txt 
!wget http://azuremlsamples.azureml.net/templatedata/PM_test.txt
!wget http://azuremlsamples.azureml.net/templatedata/PM_truth.txt

--2019-08-22 16:14:52--  http://azuremlsamples.azureml.net/templatedata/PM_train.txt
Resolving azuremlsamples.azureml.net (azuremlsamples.azureml.net)... 13.107.246.10, 2620:1ec:bdf::10
Connecting to azuremlsamples.azureml.net (azuremlsamples.azureml.net)|13.107.246.10|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 3515356 (3.4M) [text/plain]
Saving to: ‘PM_train.txt’

PM_train.txt        100%[===================>]   3.35M  --.-KB/s    in 0.1s    

2019-08-22 16:14:52 (31.2 MB/s) - ‘PM_train.txt’ saved [3515356/3515356]

--2019-08-22 16:14:53--  http://azuremlsamples.azureml.net/templatedata/PM_test.txt
Resolving azuremlsamples.azureml.net (azuremlsamples.azureml.net)... 13.107.246.10, 2620:1ec:bdf::10
Connecting to azuremlsamples.azureml.net (azuremlsamples.azureml.net)|13.107.246.10|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 2228855 (2.1M) [text/plain]
Saving to: ‘PM_test.txt’

PM_test.txt         100%[===================>]   2.

Examinemos un poco los datos con los cuales vamos a trabajar:

In [10]:
train_df = pd.read_csv('PM_train.txt', sep=" ", header=None)
train_df.head(10) # Por defecto el numero de filas sera 5.


0   1       2       3      4       5   ...    22     23     24       25  26  27
0   1   1 -0.0007 -0.0004  100.0  518.67  ...  2388  100.0  39.06  23.4190 NaN NaN
1   1   2  0.0019 -0.0003  100.0  518.67  ...  2388  100.0  39.00  23.4236 NaN NaN
2   1   3 -0.0043  0.0003  100.0  518.67  ...  2388  100.0  38.95  23.3442 NaN NaN
3   1   4  0.0007  0.0000  100.0  518.67  ...  2388  100.0  38.88  23.3739 NaN NaN
4   1   5 -0.0019 -0.0002  100.0  518.67  ...  2388  100.0  38.90  23.4044 NaN NaN
5   1   6 -0.0043 -0.0001  100.0  518.67  ...  2388  100.0  38.98  23.3669 NaN NaN
6   1   7  0.0010  0.0001  100.0  518.67  ...  2388  100.0  39.10  23.3774 NaN NaN
7   1   8 -0.0034  0.0003  100.0  518.67  ...  2388  100.0  38.97  23.3106 NaN NaN
8   1   9  0.0008  0.0001  100.0  518.67  ...  2388  100.0  39.05  23.4066 NaN NaN
9   1  10 -0.0033  0.0001  100.0  518.67  ...  2388  100.0  38.95  23.4694 NaN NaN

[10 rows x 28 columns]

Podemos observar como las dos ultimas columnas no tienen datos. Vamos a usar la funcion de pandas DROP para eliminar esas columnas.

In [11]:
train_df.drop(train_df.columns[[26,27]], axis=1, inplace=True)
train_df.head()


0   1       2       3      4   ...   21    22     23     24       25
0   1   1 -0.0007 -0.0004  100.0  ...  392  2388  100.0  39.06  23.4190
1   1   2  0.0019 -0.0003  100.0  ...  392  2388  100.0  39.00  23.4236
2   1   3 -0.0043  0.0003  100.0  ...  390  2388  100.0  38.95  23.3442
3   1   4  0.0007  0.0000  100.0  ...  392  2388  100.0  38.88  23.3739
4   1   5 -0.0019 -0.0002  100.0  ...  393  2388  100.0  38.90  23.4044

[5 rows x 26 columns]

Dado que el dataset no contiene etiquetas para las  diferentes columnas, vamos a hacerlo tomando en cuenta el [template](https://gallery.azure.ai/Experiment/Predictive-Maintenance-Step-1-of-3-data-preparation-and-feature-engineering-2) de la pagina de Azure.

In [12]:
train_df.columns = ['id', 'cycle', 'setting1', 'setting2', 'setting3', 's1', 's2', 's3',
                     's4', 's5', 's6', 's7', 's8', 's9', 's10', 's11', 's12', 's13', 's14',
                     's15', 's16', 's17', 's18', 's19', 's20', 's21']
train_df.head(10)

id  cycle  setting1  setting2  setting3  ...  s17   s18    s19    s20      s21
0   1      1   -0.0007   -0.0004     100.0  ...  392  2388  100.0  39.06  23.4190
1   1      2    0.0019   -0.0003     100.0  ...  392  2388  100.0  39.00  23.4236
2   1      3   -0.0043    0.0003     100.0  ...  390  2388  100.0  38.95  23.3442
3   1      4    0.0007    0.0000     100.0  ...  392  2388  100.0  38.88  23.3739
4   1      5   -0.0019   -0.0002     100.0  ...  393  2388  100.0  38.90  23.4044
5   1      6   -0.0043   -0.0001     100.0  ...  391  2388  100.0  38.98  23.3669
6   1      7    0.0010    0.0001     100.0  ...  392  2388  100.0  39.10  23.3774
7   1      8   -0.0034    0.0003     100.0  ...  391  2388  100.0  38.97  23.3106
8   1      9    0.0008    0.0001     100.0  ...  392  2388  100.0  39.05  23.4066
9   1     10   -0.0033    0.0001     100.0  ...  393  2388  100.0  38.95  23.4694

[10 rows x 26 columns]

Hagamos lo mismo para los datos de prueba.

In [20]:
test_df = pd.read_csv('PM_test.txt', sep=" ", header=None)
test_df.drop(test_df.columns[[26, 27]], axis=1, inplace=True)
test_df.columns = ['id', 'cycle', 'setting1', 'setting2', 'setting3', 's1', 's2', 's3',
                     's4', 's5', 's6', 's7', 's8', 's9', 's10', 's11', 's12', 's13', 's14',
                     's15', 's16', 's17', 's18', 's19', 's20', 's21']
test_df.tail(5)

id  cycle  setting1  setting2  ...   s18    s19    s20      s21
13091  100    194    0.0049    0.0000  ...  2388  100.0  38.65  23.1974
13092  100    195   -0.0011   -0.0001  ...  2388  100.0  38.57  23.2771
13093  100    196   -0.0006   -0.0003  ...  2388  100.0  38.62  23.2051
13094  100    197   -0.0038    0.0001  ...  2388  100.0  38.66  23.2699
13095  100    198    0.0013    0.0003  ...  2388  100.0  38.70  23.1855

[5 rows x 26 columns]

Miremos el archivo de ground truth:

In [21]:
truth_df = pd.read_csv('PM_truth.txt', sep= " ", header=None)
truth_df.head()

0   1
0  112 NaN
1   98 NaN
2   69 NaN
3   82 NaN
4   91 NaN

In [0]:
Escriba el codigo a continuación para quitar la columan sin información:

In [0]:
truth_df.drop(truth_df.columns[[1]], axis=1, inplace=True)
truth_df.head()

**PRE-PROCESAMIENTO DE DATOS**
Antes de comenzar con el pre.rpocesamiento de datos. Vamos a comprender en que consisten los datos:
En el archivo de entrenamiento, cada motor empieza con un grado diferente de uso inicial y variaciones de manufactura. Se asume que cada motor opera normalmente al comienzo de cada serie de tiempo y comienza a degradarse en algun punto durante la serie del ciclo de operación. La degradación es progresiva y crece en magnitud. Cuando un nivel de umbral es alcanzado, el motor es considerado inseguro. Asi que el ultimo ciclo se puede considerar como el punto de falla del correspondiente motor.

Como una forma de practicar y ver lo que sucede en el dataset de entrenamiento. Visualice cual sería el punto de falla para cada motor: 


In [35]:
#organizar los datos por id y por ciclos:
train_df =train_df.sort_values(['id', 'cycle'])
#train_df.head()
ultimo_ciclo = train_df.groupby('id')['cycle'].max()
ultimo_ciclo

id
1      192
2      287
3      179
4      189
5      269
6      188
7      259
8      150
9      201
10     222
11     240
12     170
13     163
14     180
15     207
16     209
17     276
18     195
19     158
20     234
21     195
22     202
23     168
24     147
25     230
26     199
27     156
28     165
29     163
30     194
      ... 
71     208
72     213
73     213
74     166
75     229
76     210
77     154
78     231
79     199
80     185
81     240
82     214
83     293
84     267
85     188
86     278
87     178
88     213
89     217
90     154
91     135
92     341
93     155
94     258
95     283
96     336
97     202
98     156
99     185
100    200
Name: cycle, Length: 100, dtype: int64

Como pudo notar existen 100 motores y por ejemplo el motor 1 fallará en el ciclo 192 y el motor 100 fallará en el ciclo 200.
Para el dataset de prueba, el esquema es identico al dataset de entrenamiento, solo que en este caso el ultimo ciclo no representa el punto de falla para un motor en particular.

El dataset ground truth tendrá entonces  el ciclo donde fallará cada motor del dataset de prueba.

Como un ejercicio demuestre que para el motor 2 el ultimo ciclo en el dataset de prueba es el 49 y el ciclo en que sera considerado como falla será en el 98: